File: app/src/main/java/com/fitbudget/app/domain/WaterCalculator.kt
package com.fitbudget.app.domain

import kotlin.math.ceil
import kotlin.math.roundToInt

data class WaterSnapshot(
    val consumedMl: Int,
    val targetMl: Int
) {
    val remainingMl: Int get() = (targetMl - consumedMl).coerceAtLeast(0)
    val isGoalMet: Boolean get() = targetMl > 0 && consumedMl >= targetMl
    val fraction: Double
        get() = if (targetMl <= 0) 0.0 else (consumedMl.toDouble() / targetMl).coerceIn(0.0, 1.0)
    val percent: Int get() = (fraction * 100).roundToInt()
    /** How many 250 ml glasses are still needed. */
    val glassesRemaining: Int get() = ceil(remainingMl / 250.0).toInt()
}

object WaterCalculator {

    val QUICK_AMOUNTS_ML = listOf(250, 500, 750, 1000)

    const val DEFAULT_TARGET_ML = 2500

    fun snapshot(consumedMl: Int, targetMl: Int) = WaterSnapshot(
        consumedMl = consumedMl.coerceAtLeast(0),
        targetMl = targetMl.coerceAtLeast(0)
    )

    fun total(amountsMl: List<Int>): Int = amountsMl.filter { it > 0 }.sum()

    /** Average daily intake across the supplied days (days with no log count as 0). */
    fun averageMl(dailyTotals: List<Int>): Int =
        if (dailyTotals.isEmpty()) 0 else dailyTotals.sum() / dailyTotals.size

    /** Share of days where the target was met, 0..1. */
    fun adherenceFraction(dailyTotals: List<Int>, targetMl: Int): Double {
        if (dailyTotals.isEmpty() || targetMl <= 0) return 0.0
        return dailyTotals.count { it >= targetMl }.toDouble() / dailyTotals.size
    }
}

File: ios/Sources/FitBudgetCore/Domain/WaterCalculator.swift
import Foundation

public struct WaterSnapshot: Equatable, Sendable {
    public let consumedMl: Int
    public let targetMl: Int

    public init(consumedMl: Int, targetMl: Int) {
        self.consumedMl = consumedMl
        self.targetMl = targetMl
    }

    public var remainingMl: Int { max(targetMl - consumedMl, 0) }
    public var isGoalMet: Bool { targetMl > 0 && consumedMl >= targetMl }
    public var fraction: Double {
        guard targetMl > 0 else { return 0 }
        return min(max(Double(consumedMl) / Double(targetMl), 0), 1)
    }
    public var percent: Int { Int((fraction * 100).rounded()) }
    /// How many 250 ml glasses are still needed.
    public var glassesRemaining: Int { Int((Double(remainingMl) / 250).rounded(.up)) }
}

public enum WaterCalculator {

    public static let quickAmountsMl = [250, 500, 750, 1_000]

    public static let defaultTargetMl = 2_500

    public static func snapshot(consumedMl: Int, targetMl: Int) -> WaterSnapshot {
        WaterSnapshot(consumedMl: max(consumedMl, 0), targetMl: max(targetMl, 0))
    }

    public static func total(_ amountsMl: [Int]) -> Int {
        amountsMl.filter { $0 > 0 }.reduce(0, +)
    }

    /// Average daily intake across the supplied days (days with no log count as 0).
    public static func averageMl(_ dailyTotals: [Int]) -> Int {
        guard !dailyTotals.isEmpty else { return 0 }
        return dailyTotals.reduce(0, +) / dailyTotals.count
    }

    /// Share of days where the target was met, 0...1.
    public static func adherenceFraction(dailyTotals: [Int], targetMl: Int) -> Double {
        guard !dailyTotals.isEmpty, targetMl > 0 else { return 0 }
        return Double(dailyTotals.filter { $0 >= targetMl }.count) / Double(dailyTotals.count)
    }
}
